vuepress/docs/docs/optimization.md

```diff
@@ -0,0 +1,102 @@
+
+Let's consider `getUserNameById` from the previous example. We need to define a corresponding request type that extends `zquery.Request` for a given response type:
+```scala
+case class GetUserName(id: Int) extends Request[String]
+```
+
+Now let's build the corresponding `DataSource`. We need to implement the following functions:
+```scala
+val UserDataSource = new DataSource.Service[Any, Throwable, GetUserName] {
+  override val identifier: String = ???
+  override def run(requests: Iterable[GetUserName]): ZIO[Any, Throwable, CompletedRequestMap] = ???
+}
+```
+
+We will use "UserDataSource" as our identifier. This name should not be reused for other data sources.
+```scala
+override val identifier: String = "UserDataSource"
+```
+
+We will define two different behaviors depending on whether we receive a single request or multiple requests at once.
+```scala
+override def run(requests: Iterable[GetUserName]): ZIO[Any, Throwable, CompletedRequestMap] = {
+  val resultMap = CompletedRequestMap.empty
+  requests.toList match {
+    case request :: Nil =>
+      // get user by ID e.g. SELECT name FROM users WHERE id = $id
+      val result: Task[String] = ???
+      result.map(resultMap.insert(request))
+    case batch =>
+      // get multiple users at once e.g. SELECT id, name FROM users WHERE id IN ($ids)
+      val result: Task[List[(Int, String)]] = ???
+      result.map(_.foldLeft(resultMap) { case (map, (id, name)) => map.insert(GetUserName(id))(name) })
+  }
+}
+```
+
+Now to build a `ZQuery` from it, we can use `ZQuery.fromRequestWith` and just pass the request and the data source:
+```scala
+def getUserNameById(id: Int): ZQuery[Any, Throwable, String] =
+  ZQuery.fromRequestWith(GetUserName(id))(UserDataSource)
+```
+
+To run a `ZQuery`, simply use `ZQuery#run` which will return a `ZIO[R, E, A]`.
+
+## ZQuery constructors and operators
+There are several ways to create a `ZQuery`. We've seen `ZQuery.fromRequestWith`, but you can also:
+- create from a pure value with `ZQuery.succeed`
+- create from an effect value with `ZQuery.fromEffect`
+- create from multiple queries with `ZQuery.collectAllPar` and `ZQuery.foreachPar` and their sequential equivalents `ZQuery.collectAll` and `ZQuery.foreach`
+
+If you have a `ZQuery` object, you can use:
+- `map` and `mapError` to modify the returned result or error
+- `flatMap` or `zip` to combine it with other `ZQuery` objects
+- `provide` and `provideSome` to eliminate some of the R requirements
+
+There are several ways to run a `ZQuery`:
+- `runCache` runs the query using a given pre-populated cache. This can be useful for deterministically "replaying" a query without executing any new requests.
+- `runLog` runs the query and returns its result along with the cache containing a complete log of all requests executed and their results. This can be useful for logging or analysis of query execution.
+- `run` runs the query and returns its result.
```

Comment:
Did you consider having just one methor `run` and using cache and logging based on configuration from Environment? Logging result could get logged to some environmental log.

In this way you could change the logging and caching behavior without changing the code...

Reply:
Discussion: https://discordapp.com/channels/629491597070827530/633200096393166868/638710065234182147